librz/hash/algorithms/md5/md5.c

```diff
@@ -1,242 +1,240 @@
-		}
-		index = 0;
-	} else {
-		i = 0;
-	}
-	// remaining input
-	memmove((void *)&context->buffer[index], (void *)&input[i], inputLen - i);
+	state[0] += a;
+	state[1] += b;
+	state[2] += c;
+	state[3] += d;
 }
 
-void MD5_Final(ut8 *digest, MD5_CTX *context) {
-	ut8 bits[8];
+/*
+ * Start MD5 accumulation.  Set bit count to 0 and buffer to mysterious
+ * initialization constants.
+ */
+RZ_IPI void rz_MD5Init(rz_MD5_CTX *ctx) {
+	ctx->count[0] = 0;
+	ctx->count[1] = 0;
+	ctx->state[0] = 0x67452301;
+	ctx->state[1] = 0xefcdab89;
+	ctx->state[2] = 0x98badcfe;
+	ctx->state[3] = 0x10325476;
+}
 
-	/* Save number of bits */
-	Encode(bits, context->count, 8);
+/*
+ * Update context to reflect the concatenation of another buffer full
+ * of bytes.
+ */
+RZ_IPI void rz_MD5Update(rz_MD5_CTX *ctx, const unsigned char *input, size_t len) {
+	size_t have, need;
+
+	/* Check how many bytes we already have and how many more we need. */
+	have = (size_t)((ctx->count[0] >> 3) & (MD5_BLOCK_LENGTH - 1));
+	need = MD5_BLOCK_LENGTH - have;
+
+	/* Update bitcount */
+	/*	ctx->count += (uint64_t)len << 3;*/
+	if ((ctx->count[0] += ((ut32)len << 3)) < (ut32)len) {
+		/* Overflowed ctx->count[0] */
+		ctx->count[1]++;
+	}
+	ctx->count[1] += ((ut32)len >> 29);
+
+	if (len >= need) {
+		if (have != 0) {
+			memcpy(ctx->buffer + have, input, need);
+			MD5Transform(ctx->state, ctx->buffer);
+			input += need;
+			len -= need;
+			have = 0;
+		}
 
-	/* Pad out to 56 mod 64.  */
-	ut32 index = (ut32)((context->count[0] >> 3) & 0x3f);
-	ut32 padLen = (index < 56) ? (56 - index) : (120 - index);
-	MD5_Update(context, PADDING, padLen);
+		/* Process data in MD5_BLOCK_LENGTH-byte chunks. */
+		while (len >= MD5_BLOCK_LENGTH) {
+			MD5Transform(ctx->state, input);
+			input += MD5_BLOCK_LENGTH;
+			len -= MD5_BLOCK_LENGTH;
+		}
+	}
 
-	/* Append length (before padding) */
-	MD5_Update(context, bits, 8);
+	/* Handle any remaining bytes of data. */
+	if (len != 0) {
+		memcpy(ctx->buffer + have, input, len);
+	}
+}
 
-	/* Store state in digest */
-	Encode(digest, context->state, 16);
+/*
+ * Final wrapup - pad to 64-byte boundary with the bit pattern
+ * 1 0* (64-bit count of bits processed, MSB-first)
+ */
+RZ_IPI void rz_MD5Final(unsigned char digest[MD5_DIGEST_LENGTH], rz_MD5_CTX *ctx) {
+	ut8 count[8];
+	size_t padlen;
+	int i;
+
+	/* Convert count to 8 bytes in little endian order. */
+	PUT_64BIT_LE(count, ctx->count);
+
+	/* Pad out to 56 mod 64. */
+	padlen = MD5_BLOCK_LENGTH -
+		((ctx->count[0] >> 3) & (MD5_BLOCK_LENGTH - 1));
+	if (padlen < 1 + 8) {
+		padlen += MD5_BLOCK_LENGTH;
+	}
+	rz_MD5Update(ctx, PADDING, padlen - 8); /* padlen - 8 <= 64 */
+	rz_MD5Update(ctx, count, 8);
 
-	/* Zeroize sensitive information.  */
-	rz_mem_memzero((void *)context, sizeof(*context));
+	if (digest != NULL) {
+		for (i = 0; i < 4; i++) {
+			rz_write_le32(digest + i * 4, ctx->state[i]);
+		}
+	}
+	bzero(ctx, sizeof(*ctx)); /* in case it's sensitive */
```

Comment:
i'm not 100% sure that bzero is portable. (https://stackoverflow.com/questions/17096990/why-use-bzero-over-memset/17097088) so i would suggest to change this to memset

Reply:
Good catch, let's use `rz_mem_memzero`